src/Executor/Executor.php

```diff
@@ -222,12 +223,15 @@ private static function getOperationRootType(Schema $schema, OperationDefinition
     private static function executeFieldsSerially(ExecutionContext $exeContext, ObjectType $parentType, $sourceValue, $path, $fields)
     {
         $results = [];
+        $sourceValue = self::completePromiseIfNeeded($sourceValue);
+
         foreach ($fields as $responseName => $fieldASTs) {
             $fieldPath = $path;
             $fieldPath[] = $responseName;
             $result = self::resolveField($exeContext, $parentType, $sourceValue, $fieldASTs, $fieldPath);
 
             if ($result !== self::$UNDEFINED) {
+                $result = self::completePromiseIfNeeded($result);
```

Comment:
that's where the chaining promise enters in action! let's say we have this query:

```js
{
  user1: user(id: 1) { name friends { name }}
  user2: user(id: 2) { name friends { name }}
}
```

user resolver:
```php
function ($id) {
  return $this->userDataLoader->load($id);
}
```

friends resolver:
```php
function ($value) {
  return $value->then(function($values)) {
    return $this->userDataLoader->loadMany($ids);
  } 
}
```

`wait` will call user resolver that return promise for user 1 that is resolve by the data loader, resolving promise for user 1 will also resolve user 2 promise (thanks to dataLoader). friends for user 1 will be resolve and same for friends for user 2. only 2 requests to db against 2 + (x * friends user 1) +  (n * friends user 2)...

It is difficult to show without a concrete example, thats what i'm working on...